orm/src/main/java/org/hibernate/search/hcore/impl/SearchFactoryReference.java

```diff
@@ -22,16 +20,16 @@
  */
 public final class SearchFactoryReference implements Service {
 
-	private volatile CompletableFuture<ExtendedSearchIntegrator> extendedIntegratorFuture;
+	private volatile ExtendedSearchIntegrator extendedIntegrator;
 
-	public void initialize(CompletableFuture<ExtendedSearchIntegrator> future) {
-		extendedIntegratorFuture = future;
+	public void initialize(ExtendedSearchIntegrator extendedIntegrator) {
+		this.extendedIntegrator = extendedIntegrator;
 	}
 
 	public ExtendedSearchIntegrator getSearchIntegrator() {
-		final CompletableFuture<ExtendedSearchIntegrator> value = this.extendedIntegratorFuture;
+		ExtendedSearchIntegrator value = this.extendedIntegrator;
 		if ( value != null ) {
-			return value.join();
+			return value;
```

Comment:
Sounds like you're assume that the two frameworks (Hibernate, CDI) are booting in two different threads running fully in parallel, and in parallel with the end user's app. Is that the case?

With the added synchronizations I've been addressing _visibility_ concerns, in case the two frameworks are initialized _by_ different threads - potentially something some containers will do - but I don't expect the application to be considered "deployed" until all frameworks have been initialized.

Let me try explain that in other words: I've addressed cross-thread visibility concerns, not timing. The "happens-before" rules should apply and I think we'll need to have faith in the containers?